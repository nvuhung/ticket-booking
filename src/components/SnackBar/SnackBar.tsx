import React from 'react';
import { useSelector } from 'react-redux';
import styled, { keyframes, css } from 'styled-components';

import { ISnackbarState } from '../../store/snackbar/reducers';
import { theme } from '../../styles/theme';

const SnackBar: React.FC = () => {
  const { isShow, message, type, timing } = useSelector(
    (state: { snackbar: ISnackbarState }) => state.snackbar
  );

  return isShow ? (
    <SnackBarStyled delay={timing - 0.5} backgroundColor={theme.colors[type]} show={isShow}>
      {message}
    </SnackBarStyled>
  ) : null;
};

const fadeIn = keyframes`
  from {bottom: 0; opacity: 0;}
  to {bottom: 30px; opacity: 1;}
`;

const fadeOut = keyframes`
  from {bottom: 30px; opacity: 1;}
  to {bottom: 0; opacity: 0;}
`;

const SnackBarStyled = styled.div<{ show: boolean; backgroundColor: string; delay: number }>`
  min-width: 250px;
  margin-left: -125px;
  background-color: ${props => props.backgroundColor};
  color: ${theme.colors.white};
  text-align: center;
  border-radius: 2px;
  padding: 16px;
  position: fixed;
  z-index: 1;
  left: 50%;
  bottom: 30px;

  ${props =>
    props.show &&
    css`
      animation: ${fadeIn} 0.5s, ${fadeOut} 0.5s ${props.delay}s;
    `}
`;

export default React.memo(SnackBar);
