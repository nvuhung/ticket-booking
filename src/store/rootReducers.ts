import { combineReducers } from 'redux';

import seatReducer from './seat/reducers';
import snackbarReducer from './snackbar/reducers';

export const rootReducers = combineReducers({
  seat: seatReducer,
  snackbar: snackbarReducer
});
