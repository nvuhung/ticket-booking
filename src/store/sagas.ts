import { all, fork } from 'redux-saga/effects';

import { saga as seatSaga } from './seat/sagas';
import { saga as snackbarSaga } from './snackbar/sagas';

export default function* rootSaga(): Generator {
  yield all([fork(seatSaga), fork(snackbarSaga)]);
}
