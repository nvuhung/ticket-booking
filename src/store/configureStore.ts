import { applyMiddleware, compose, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';

import { rootReducers } from './rootReducers';
import rootSaga from './sagas';

// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const sagaMiddleware = createSagaMiddleware();
const enhancers = [applyMiddleware(sagaMiddleware)];

const store = createStore(rootReducers, composeEnhancers(...enhancers));
sagaMiddleware.run(rootSaga);

export default { ...store };
