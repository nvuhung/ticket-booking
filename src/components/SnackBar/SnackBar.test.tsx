import React from 'react';
import 'jest-styled-components';

import SnackBar from './SnackBar';
import { mount, shallow } from '../../test/enzyme';
import store from '../../store/configureStore';
import { Provider } from 'react-redux';

describe('Snackbar', () => {
  let mountedComp: any;

  const wrapper = (bShallow: boolean) => {
    if (!mountedComp) {
      mountedComp = bShallow
        ? shallow(
            <Provider store={store}>
              <SnackBar />
            </Provider>
          )
        : mount(
            <Provider store={store}>
              <SnackBar />
            </Provider>
          );
    }
    return mountedComp;
  };

  it('renders without crashing', () => {
    const comp = wrapper(false);
    expect(comp).toMatchSnapshot();
  });
});
