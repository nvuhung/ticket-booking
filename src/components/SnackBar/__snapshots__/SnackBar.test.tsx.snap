// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar renders without crashing 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Memo(SnackBar) />
</Provider>
`;
